InformationScripting/src/queries/AstQuery.cpp

```diff
@@ -0,0 +1,252 @@
+/***********************************************************************************************************************
+**
+** Copyright (c) 2011, 2015 ETH Zurich
+** All rights reserved.
+**
+** Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
+** following conditions are met:
+**
+**    * Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+**      disclaimer.
+**    * Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+**      following disclaimer in the documentation and/or other materials provided with the distribution.
+**    * Neither the name of the ETH Zurich nor the names of its contributors may be used to endorse or promote products
+**      derived from this software without specific prior written permission.
+**
+**
+** THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+** INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+** DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+** SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+** SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+** WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+** OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+**
+***********************************************************************************************************************/
+
+#include "AstQuery.h"
+
+#include "OOModel/src/declarations/Class.h"
+#include "OOModel/src/declarations/Project.h"
+
+#include "../graph/Graph.h"
+#include "../graph/InformationNode.h"
+
+#include "../visitors/AllNodesOfType.h"
+
+namespace InformationScripting {
+
+AstQuery::AstQuery(QueryType type, Model::Node* target, QStringList args)
+	: target_{target}, type_{type}
+{
+	if (args.size())
+	{
+		if (args[0] == "g") scope_ = Scope::Global;
+		else if (args[0] == "of") scope_ = Scope::Input;
+	}
+}
+
+QList<Graph*> AstQuery::execute(QList<Graph*> input)
+{
+	QList<Graph*> result;
+	switch (type_)
+	{
+		case QueryType::Classes:
+			result = {classesQuery(input)};
+			break;
+		case QueryType::Methods:
+			result = {methodsQuery(input)};
+			break;
+		case QueryType::BaseClasses:
+			result = {baseClassesQuery(input)};
+			break;
+		case QueryType::ToClass:
+			result = {toClassNode(input)};
+			break;
+		case QueryType::CallGraph:
+			result = {callGraph(input)};
+			break;
+	}
```

Comment:
Just as a side note clang does this at compile time, i.e. it warns for each unhandled case.